socorro/tests/test_load_processed_crashes_into_es.py

```diff
@@ -0,0 +1,132 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+from datetime import date, datetime, time, timezone
+import json
+import pytest
+
+from click.testing import CliRunner
+
+from bin.load_processed_crashes_into_es import load_crashes
+from socorro.lib.libooid import create_new_ooid
+from socorro import settings
+
+
+# Setup helper, so we have some processed crashes to load into ES
+def load_crashes_into_crashstorage_source(helper, date_str="2024-05-01", num_crashes=2):
+    bucket = settings.STORAGE["options"]["bucket"]
+    helper.create_bucket(bucket)
+
+    expected_processed_crashes = []
+    date_args = [int(date_part) for date_part in date_str.split("-")]
+    date_datetime = date(*date_args)
+    expected_processed_crash_base = {
+        "date_processed": datetime.combine(
+            date_datetime, time.min, timezone.utc
+        ).isoformat(),
+    }
+    raw_crash_base = {"submitted_timestamp": date_str}
+
+    for _ in range(num_crashes):
+        # Encode the date in the crash ID, so we can determine the correct
+        # Elasticsearch index by crash ID downstream.
+        crash_id = create_new_ooid(timestamp=date_datetime)
+        processed_crash = {**expected_processed_crash_base, "uuid": crash_id}
+        expected_processed_crashes.append(processed_crash)
+
+        # Upload raw crash for this crash ID, since only the raw crash
+        # path contains the date for lookups.
+        raw_crash = {**raw_crash_base, "uuid": crash_id}
+        date_str_fmt = raw_crash["submitted_timestamp"].replace("-", "")
+        helper.upload(
+            bucket_name=bucket,
+            key=f"v1/raw_crash/{date_str_fmt}/{crash_id}",
+            data=json.dumps(raw_crash).encode("utf-8"),
+        )
+
+        # Upload processed crash for the crash ID
+        helper.upload(
+            bucket_name=bucket,
+            key=f"v1/processed_crash/{crash_id}",
+            data=json.dumps(processed_crash).encode("utf-8"),
+        )
+
+    return expected_processed_crashes
+
+
+def test_it_runs():
+    """Test whether the module loads and spits out help."""
+    runner = CliRunner()
+    result = runner.invoke(load_crashes, ["--help"])
+    assert result.exit_code == 0
+
+
+@pytest.mark.skipif(
+    settings.CLOUD_PROVIDER != "AWS",
+    reason="Skipping test because CLOUD_PROVIDER is not set to 'AWS'",
+)
+def test_it_loads_processed_crashes_by_date_s3(s3_helper, es_helper):
```

Comment:
This test and the corresponding GCS one testing the date argument are quite slow. Since this is testing a one-off GCP migration script, we should consider some options for speeding things up.

One option is to skip the date tests in CI.

Another option is to add test-only config to the `load_crashes` command to use only the date in the raw_crash prefix when searching for raw crashes by date, since the majority of the slowness comes from including all possible combinations of the first three characters of the crash ID in the prefix.

Perhaps there are other options (other than do nothing!)?

Reply:
any reason we can't drop the crash-prefix splitting entirely?

Reply:
or make it an option that gets passed into the script with variable length, and then in tests we would submit an empty prefix to do effectively disable the prefix splitting?

Reply:
> any reason we can't drop the crash-prefix splitting entirely?

That's a really good question.

I assume the reason we do this, is because in prod, we expect a "large" volume of crashes on any given day, so I tried to dig up some data:
* Using the `supersearch` CLI, I looked up a few dates, and we get between 80-100k crashes for a 24 hour period.
* All other things being equal, that means there are about 25 processed crashes (i.e. unique crash IDs) for each unique three-character crash ID prefix[1].

But, as you've noted elsewhere, not only do we chunk the search prefix, we also get paginated results back from the crashstorage source AND we split all of these checks across multiple workers (4 by default). So my thinking is that no, we don't need the crash-prefix splitting on the surface of things, and it would be simplest/fastest to remove it entirely from the script.

However, this [`firstchars` prefix splitting](https://github.com/mozilla-services/socorro/blob/00d064a90eaf533fda9a4ea0c3b1dddf10781c78/webapp/crashstats/crashstats/management/commands/verifyprocessed.py#L85) is based off of the `check_crashids_for_date` function that's part of an existing `verifyprocessed` Django command which verifies that a raw crash has been processed and saved to all three crashstorage locations (S3, S3 telemetry and ES in this case). @willkg's: To your recollection, why was this done, and what do you think is best here?

[1] Since the first 3 chars can be any of 16 hexadecimal characters, that's 16^3 or 4096 possible combinations. If crash IDs are randomly assigned, and we assume a uniform distribution, that gives 100k/4096 or 25 unique crash IDs per combination.

Reply:
This is how we split the work across multiple workers in `verifyprocessed`. The combinations are added to a queue and the workers pull a combination from the queue, get a listing, then run verification for that crash id.

https://github.com/mozilla-services/socorro/blob/00d064a90eaf533fda9a4ea0c3b1dddf10781c78/webapp/crashstats/crashstats/management/commands/verifyprocessed.py#L132-L150

Whether we want to do that same thing in this script depends on how it's structured in regards to breaking up the work across multiple workers.

Reply:
Okay thanks. I'm going to defer making any changes to `verifyprocessed` until we try this script out for the migration. I've gone in a different direction in terms of what work is being done in parallel, but I think either way works, and which is "better" depends on what our crashstorage data looks like in prod. This might be in the realm of premature optimization, so I'll table the discussion for now.